tfhe/src/integer/server_key/radix_parallel/add.rs

```diff
@@ -139,8 +153,335 @@ impl ServerKey {
             }
         };
 
-        self.unchecked_add_assign(lhs, rhs);
-        self.full_propagate_parallelized(lhs);
+        if self.is_eligible_for_parallel_carryless_add(lhs.blocks.len()) {
+            self.unchecked_add_assign_parallelized_low_latency(lhs, rhs, AddExtraOne::No);
+        } else {
+            self.unchecked_add_assign(lhs, rhs);
+            self.full_propagate_parallelized(lhs);
+        }
+    }
+
+    pub fn add_parallelized_work_efficient<PBSOrder: PBSOrderMarker>(
+        &self,
+        ct_left: &RadixCiphertext<PBSOrder>,
+        ct_right: &RadixCiphertext<PBSOrder>,
+    ) -> RadixCiphertext<PBSOrder> {
+        let mut ct_res = ct_left.clone();
+        self.add_assign_parallelized_work_efficient(&mut ct_res, ct_right);
+        ct_res
+    }
+
+    pub fn add_assign_parallelized_work_efficient<PBSOrder: PBSOrderMarker>(
+        &self,
+        ct_left: &mut RadixCiphertext<PBSOrder>,
+        ct_right: &RadixCiphertext<PBSOrder>,
+    ) {
+        let mut tmp_rhs: RadixCiphertext<PBSOrder>;
+
+        let (lhs, rhs) = match (
+            ct_left.block_carries_are_empty(),
+            ct_right.block_carries_are_empty(),
+        ) {
+            (true, true) => (ct_left, ct_right),
+            (true, false) => {
+                tmp_rhs = ct_right.clone();
+                self.full_propagate_parallelized(&mut tmp_rhs);
+                (ct_left, &tmp_rhs)
+            }
+            (false, true) => {
+                self.full_propagate_parallelized(ct_left);
+                (ct_left, ct_right)
+            }
+            (false, false) => {
+                tmp_rhs = ct_right.clone();
+                rayon::join(
+                    || self.full_propagate_parallelized(ct_left),
+                    || self.full_propagate_parallelized(&mut tmp_rhs),
+                );
+                (ct_left, &tmp_rhs)
+            }
+        };
+
+        self.unchecked_add_assign_parallelized_low_latency(lhs, rhs, AddExtraOne::No);
+    }
+
+    pub(crate) fn is_eligible_for_parallel_carryless_add(&self, num_blocks: usize) -> bool {
+        // having 3-bits is a hard requirement
+        let total_modulus = self.key.message_modulus.0 * self.key.carry_modulus.0;
+        let has_at_least_3_bits = total_modulus >= (1 << 3);
+        // This threshold is arbitrary
+        let has_enough_threads = rayon::current_num_threads() >= (num_blocks / 2);
```

Comment:
not sure about this if it's not tied to some measurements, as the more thread heavy version may not trigger for "obscure reasons" while we want to alway trigger it in the default path (if parameters are compatible) and let users use the other version if they want to

Reply:
well it's not that arbitrary as the algorithm will at most use num_blocks / 2 threads

Reply:
If we always enable the parallel version, we are likely to slowdown computations on traditionnal laptops that have less threads, that was the idea behind that check

Reply:
I think it's worth discussing, but I understand where it comes from

Reply:
I removed the check on threads for now